fetch_worker.go

```diff
@@ -71,21 +90,45 @@ func (w *fetchWorker) run(ctx context.Context) (err error) {
 
 	defer func() { err = errors.Join(err, tx.Commit()) }()
 
-	start, end := w.start, w.end
-	for {
-		rows, err := w.selectRowsChunk(ctx, tx, start, end)
+	sdk.Logger(ctx).Info().
+		Uint64("start", w.start).
+		Uint64("end", w.end).
+		Uint64("fetchSize", w.config.fetchSize).
+		Msg("fetching rows")
+
+	for chunkStart := w.start; chunkStart <= w.end; chunkStart += w.config.fetchSize {
+		chunkEnd := chunkStart + w.config.fetchSize
+		sdk.Logger(ctx).Info().
+			Uint64("chunk start", chunkStart).
+			// the where clause is exclusive at the end
+			Uint64("chunk end", chunkEnd-1).
+			Msg("fetching chunk")
+		rows, err := w.selectRowsChunk(ctx, tx, chunkStart, chunkEnd)
 		if err != nil {
 			return fmt.Errorf("failed to select rows chunk: %w", err)
 		}
 		if len(rows) == 0 {
-			break
+			continue
 		}
 
 		for _, row := range rows {
-			start++
+			sdk.Logger(ctx).Trace().Msgf("fetched row: %+v", row)
+
+			primaryKeyVal := row[w.config.primaryKey]
+			if primaryKeyVal == nil {
+				return ErrPrimaryKeyNotFoundInRow
+			}
+
+			lastRead, err := common.ConvertPrimaryKey(primaryKeyVal)
+			if err != nil {
+				return fmt.Errorf("failed to convert primary key: %w", err)
+			}
+
 			position := common.TablePosition{
-				LastRead:    start,
-				SnapshotEnd: end,
+				// plus 1 so that we start reading from the correct place, otherwise
+				// we would read the same record twice on multiple snapshots
+				LastRead:    lastRead + 1,
```

Comment:
:+1:

Reply:
If ids are not sequential they cannot be used for sorting rows for the snapshot, regardless of the `+1`. I'll adress #53 soon, and I guess that in my fix I will have to modify how I handle the `opencdc.Position`.

Reply:
I mean this doesn't play nicely when getting the minimum value. But after a bit more thought, a record having the position of the next row instead of his current one is strange aswell. I took a look at the postgres connector, and it also does the range exclusive to inclusive, so I might aswell do the same.

I did not do it initially because a bunch of tests did fail, which I've now fixed. Just pushed a change with the new iteration + some explanation on why I'm getting the mininum value the way I'm getting it.

Thanks for the review @nickchomey! I feel now the snapshot is a bit more understandable indeed.